### ISSUE TITLE
Simplifying the options for the credentials list in the envelope

### ISSUE THREAD

**Issue by kevinlewi** [cfrg/draft-irtf-cfrg-opaque#99]
This issue is related to #87 and #90, in which the complexities of the setting of the envelope have come up.

One problem I foresee happening with allowing the server to choose what parameters should be optional, [which go in secret_credentials, which go in cleartext_credentials, etc.](https://github.com/cfrg/draft-irtf-cfrg-opaque/blame/master/draft-irtf-cfrg-opaque.md#L438-L442) is that it becomes cumbersome to manage exhaustive test vectors that would test all of the edge cases (combinations of these settings). And if we choose to only provide test vectors for some default setting, we run the risk of an implementation in the wild incorrectly implementing support for a non-default setting that we did not provide test vectors for.

To remedy this, one approach we could take would be to allow for only a small (maybe 2?) set of options for organizing the credentials in the envelope. Perhaps a "base mode" and a "custom identifier mode", and we could spend a single byte to dictate the mode for the envelope (similar to the VOPRF RFC with the byte representing the base mode or verifiable mode). Base mode would be some default, and "custom identifier mode" would support a custom idU and idS being included as secret_credentials.

If this sounds good, then now we just need to decide what modes we will support, and what each mode will look like.

**Comment by chris-wood** [comment]
I like this idea. Given that the number of credential types is likely quite small, I'm in favor of simplifying things and offering two modes here. For the "custom identifier mode," either idU or idS could be nil, if desired, right? Or would we require that they both non-nil?

**Comment by kevinlewi** [comment]
I think we should allow for nil idU and idS, because it is not easy to check/enforce that idU and idS, and even if it were easily checkable, it is still an additional check that the OPAQUE implementation would have to do, which increases complexity.

**Comment by bytemare** [comment]
Both idu and ids are used in the "info" / "transcript" (naming is not consistent) parameter in HKDF. Having them _nillable_ means the HKDF only relies on the nonces. But I don't know if that's a problem.

**Comment by kevinlewi** [comment]
The default behavior (base mode) is for them to be set to pkU and pkS. Only in customIdentifier mode can they be set to arbitrary strings. We discussed this a bit -- and we decided to allow null values to be set, since it is up to the wrapper protocol to dictate what should represent valid identities for idU and idS. Even if we were to disallow null values, they could still be set to 1-byte values... so the best we can do is to advise for defaulting to base mode unless the implementer really wants to use the custom identifier mode in this manner.
